File: 1-100/1-two-sum.py
class Solution:
  def twoSum(self, nums: List[int], target: int) -> List[int]:
    dic = {}
    for index, num in enumerate(nums):
      # target = num + m
      m = target - num

      # return the indices of m and num if m is found in
      # the dictionary
      if m in dic:
        return [dic[m], index]
      # store num as key and its index as value in dictionary
      else:
        dic[num] = index

File: 1-100/1-two-sum.java
// O(n)
// 1 ms, faster than 99.96%
// 39.6 MB, less than 44.16%
class Solution {
  public int[] twoSum(int[] nums, int target) {
    var map = new HashMap<Integer, Integer>();
    var res = new int[2];
    for(int i=0;i<nums.length;i++) {
      var hasKey = map.containsKey(target - nums[i]);
      if(hasKey) {
        res[0] = map.get(target - nums[i]);
        res[1] = i;
        return res;
      }
      
      map.put(nums[i], i);
    }
    
    return res;
  }
}